File: app-lib/framework/app-core/src/main/scala/edu/member/persistence/Member.scala
/*
 * Copyright IxiaS, Inc. All Rights Reserved.
 *
 * For the full copyright and license information,
 * please view the LICENSE file that was distributed with this source code.
 */

package edu.member.persistence

import javax.inject.*
import scala.concurrent.Future
import ixias.db.slick.{ SlickBaseRepository, SlickDatabaseContext }
import ixias.core.persistence.HostSpec

import edu.member.persistence.table.MemberTable

/**
 * Repository for Member persistence.
 */
@Singleton
class MemberRepository @Inject()(
  table: MemberTable,
  ctx:   SlickDatabaseContext
) extends SlickBaseRepository(table, ctx):
  import api.*

  /**
   * Resolve a member by login email (used by signup/login).
   */
  def findByEmail(email: String): Future[Option[EntityEmbeddedId]] =
    RunDBAction(HostSpec.REPLICA): slick =>
      slick
        .filter(_.email === email)
        .result.headOption

File: app/web/src/lib/member.ts
/*
 * Copyright IxiaS, Inc. All Rights Reserved.
 *
 * For the full copyright and license information,
 * please view the LICENSE file that was distributed with this source code.
 */

import { api, request, requestVoid } from '@app/api';

export type Member = { id: number; uuid: string; name: string; email: string };

export type AuthResult = { ok: true } | { ok: false; error: string };

/**
 * The logged-in member, or null when not authenticated.
 *
 * A 401 is the normal "logged out" answer here rather than a failure, so it
 * collapses to null instead of surfacing a message.
 */
export async function fetchMe(): Promise<Member | null> {
  try {
    const { data, response } = await api.GET('/member/api/me', {
      credentials: 'include',
    });
    if (!response.ok || !data) return null;
    return data;
  } catch {
    return null;
  }
}

/**
 * Register and sign in in one step — the backend issues the session cookie as
 * part of the signup response, so no follow-up login is needed.
 */
export async function signup(
  email: string,
  password: string,
  name: string,
  nameKana: string,
): Promise<AuthResult> {
  return requestVoid(
    api.POST('/member/api/signup', {
      body:        { email, password, name, nameKana },
      credentials: 'include',
    }),
    '登録に失敗しました',
  );
}

/**
 * Email/password login. The backend answers an unknown address and a wrong
 * password identically, so nothing here distinguishes them either.
 */
export async function login(email: string, password: string): Promise<AuthResult> {
  return requestVoid(
    api.POST('/member/api/login', {
      body:        { email, password },
      credentials: 'include',
    }),
    'ログインに失敗しました',
  );
}

/**
 * Revoke the session server-side and drop the cookie. Idempotent — the backend
 * answers 204 even when the caller has no valid cookie.
 */
export async function logout(): Promise<AuthResult> {
  return requestVoid(
    api.POST('/member/api/logout', { credentials: 'include' }),
    'ログアウトに失敗しました',
  );
}
